File: #1671.minimum-number-of-removals-to-make-mountain-array.cpp
class Solution {
public:
    int minimumMountainRemovals(vector<int>& nums) {

        int size = nums.size();
        vector<int> v(size, 0);
        vector<int> v2(size, 0);
        for (int i = 1; i <size; i++) {
            for (int j = 0; j < i; j++) {
                if(nums[i]>nums[j]) {
                    v[i]=max(v[i],v[j]+1);
                }
            }
        }

        for (int i = size - 2; i >= 0; i--) {
            for (int j = size - 1; j > i; j--) {
                if (nums[i]>nums[j]) {
                    v2[i] = std::max(v2[i], v2[j] + 1);
                }
            }
        }

        int result = INT_MAX;
        for (int i = 1; i < size-1; i++) {
            if (v[i] == 0 || v2[i] == 0) {
                continue;
            }

            result = std::min( result, (i-v[i])+(size-i-1-v2[i]));
        }
        return result;
    }
};

File: #1671.minimum-number-of-removals-to-make-mountain-array.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var minimumMountainRemovals = function(nums) {
    let n = nums.length;
    let lisL = [...Array(n)].map(d => 1);
    for (let i = 0; i < n; i++) {
        for (let j = 0; j < i; j++) { 
            if (nums[i] > nums[j]) {
                lisL[i] = Math.max(lisL[i], lisL[j] + 1);
            }
        }
    }

    let ldsR = [...Array(n)].map(d => 1);
    for (let i = n - 1; i >= 0; i--) {
        for (let j = n - 1; j > i; j--) {
            if (nums[i] > nums[j]) {
                ldsR[i] = Math.max(ldsR[i], ldsR[j] + 1);
            }
        }
    }

    let result = Infinity;
    for (let i = 1; i < n - 1; i++) {
        let mountainLength = lisL[i] + ldsR[i] - 1;
        if (lisL[i] > 1 && ldsR[i] > 1) {
            result = Math.min(result, n - mountainLength);
        }
    }
    return result;
};
